torch/utils/checkpoint.py

```diff
@@ -662,15 +658,12 @@ def unpack_hook(holder):
 
             if holder.handles[gid] is None:
                 raise RuntimeError(
-                    "If you are calling ctx.saved_tensor in backward, make sure to do so only once. "
-                    "Otherwise please open an issue with details on your use case."
-                )
-            if holder.handles[gid] not in frame.recomputed[gid]:
-                raise RuntimeError(
-                    "Attempt to retrieve a tensor saved by autograd multiple times without checkpoint"
-                    " recomputation being triggered in between, this is not currently supported. Please"
-                    " open an issue with details on your use case."
+                    "torch.utils.checkpoint: unpack is being triggered for a tensor that was either"
+                    "never recomputed, or already unpacked once. If you are calling ctx.saved_tensor "
+                    "in backward, make sure to do so only once. Otherwise please open an issue with "
+                    "details on your use case."
                 )
+            assert holder.handles[gid] in frame.recomputed[gid], "this is a bug, please file an issue"
```

Comment:
Making this runtime error into an internal assert because we don't ever expect to trigger that runtime error.

Suppose if we were able to reach that runtime error, one of the two following must hold:
1) the handle was never an entry to frame.recomputed[gid]
2) the handle was an entry at some point but the entry was cleared

(1) is false because for handle to not be None, it must've been added to frame.recompute[gid] at some point
(2) if handle was once an entry, and then that entry is now gone - since frame.recompute[gid] is a weak key dictionary and we never explicitly remove any entries, this implies that holder.handles[gid] is no longer alive, which is false, given that the only time we set a non-None value to it is if it is None, and the only value we set to it is None.

The reason we once had this runtime error was because we previously create holder.handles[gid] in the outer pack_hook, so if recompute pass does not recompute handle, e.g. because we saved fewer tensors in the backward pass, we would've errored out here.

Reply:
Pain "assert" are not great, you can make this an AssertionError if you want though.